Implement action generic client

Barry-Xu-2018 opened the pull request:
Address https://github.com/ros-infrastructure/rep/pull/405

Action GenericClient will be used in rosbag2 to support actions.

Barry-Xu-2018 committed:
Implement action generic client

Signed-off-by: Barry Xu <barry.xu@sony.com>

ahcorde commented on rclcpp_action/include/rclcpp_action/client_base.hpp:
```diff
@@ -0,0 +1,319 @@
+
+  /// \internal
+  RCLCPP_ACTION_PUBLIC
+  size_t
```
include `<cstddef>`

ahcorde commented on rclcpp_action/include/rclcpp_action/client_base.hpp:
```diff
@@ -0,0 +1,319 @@
+  action_server_is_ready() const;
+
+  /// Wait for action_server_is_ready() to become true, or until the given timeout is reached.
+  template<typename RepT = int64_t, typename RatioT = std::milli>
```
include `<cstdint>`

ahcorde commented on rclcpp_action/include/rclcpp_action/client_base.hpp:
```diff
@@ -0,0 +1,319 @@
+   */
+  RCLCPP_ACTION_PUBLIC
+  void
+  set_on_ready_callback(std::function<void(size_t, int)> callback) override;
```
include `<functional>`

ahcorde commented on rclcpp_action/include/rclcpp_action/client_base.hpp:
```diff
@@ -0,0 +1,319 @@
+    const void * user_data);
+
+  // Mutex to protect the callbacks storage.
+  std::recursive_mutex listener_mutex_;
```
include `<mutex>`

ahcorde commented on rclcpp_action/include/rclcpp_action/create_generic_client.hpp:
```diff
@@ -0,0 +1,80 @@
+ *   If `nullptr`, then the action client is added to the default callback group.
+ * \param[in] options Options to pass to the underlying `rcl_action_client_t`.
+ */
+typename GenericClient::SharedPtr
```
document return

ahcorde commented on rclcpp_action/include/rclcpp_action/create_generic_client.hpp:
```diff
@@ -0,0 +1,80 @@
+ *   If `nullptr`, then the action client is added to the default callback group.
+ * \param[in] options Options to pass to the underlying `rcl_action_client_t`.
+ */
+template<typename NodeT>
```
document return

ahcorde commented on rclcpp_action/include/rclcpp_action/generic_client.hpp:
```diff
@@ -0,0 +1,319 @@
+  using CancelRequest = typename action_msgs::srv::CancelGoal_Request;
+  using CancelResponse = typename action_msgs::srv::CancelGoal_Response;
+  using WrappedResult = typename GenericClientGoalHandle::WrappedResult;
+  using GoalResponseCallback = std::function<void (typename GenericClientGoalHandle::SharedPtr)>;
```
include `<functional>`

ahcorde commented on rclcpp_action/include/rclcpp_action/generic_client.hpp:
```diff
@@ -0,0 +1,319 @@
+  std::shared_future<typename GenericClientGoalHandle::SharedPtr>
+  async_send_goal(
+    const Goal goal,
+    size_t goal_size,
```
include `<cstddef>`

ahcorde commented on rclcpp_action/include/rclcpp_action/generic_client.hpp:
```diff
@@ -0,0 +1,319 @@
+
+  /// Send an action goal request and asynchronously get the result.
+  /**
+   * \param[in] goal The goal request (uuid+goal).
```
```suggestion
   * \param[in] goal_request The goal request (uuid+goal).
```

ahcorde commented on rclcpp_action/include/rclcpp_action/generic_client.hpp:
```diff
@@ -0,0 +1,319 @@
+   * <a href="https://github.com/ros2/rcl_interfaces/blob/master/action_msgs/srv/CancelGoal.srv">
+   * action_msgs/CancelGoal.srv</a>.
+   */
+  std::shared_future<typename CancelResponse::SharedPtr>
```
include `<future>`

ahcorde commented on rclcpp_action/include/rclcpp_action/generic_client.hpp:
```diff
@@ -0,0 +1,319 @@
+  IntrospectionMessageMembersPtr feedback_type_members_;
+
+  std::map<GoalUUID, typename GenericClientGoalHandle::WeakPtr> goal_handles_;
+  std::recursive_mutex goal_handles_mutex_;
```
include `<mutex>`

ahcorde commented on rclcpp_action/include/rclcpp_action/generic_client_goal_handle.hpp:
```diff
@@ -0,0 +1,154 @@
+  };
+
+  using FeedbackCallback =
+    std::function<void (
```
include `<functional>`

ahcorde commented on rclcpp_action/include/rclcpp_action/generic_client_goal_handle.hpp:
```diff
@@ -0,0 +1,154 @@
+  get_goal_stamp() const;
+
+  /// Get the goal status code.
+  int8_t
```
include `<cstdint>`

ahcorde commented on rclcpp_action/src/create_generic_client.cpp:
```diff
@@ -0,0 +1,84 @@
+  rclcpp::CallbackGroup::SharedPtr group,
+  const rcl_action_client_options_t & options)
+{
+  std::weak_ptr<rclcpp::node_interfaces::NodeWaitablesInterface> weak_node =
```
include `<memory>`

ahcorde commented on rclcpp_action/src/generic_client.cpp:
```diff
@@ -0,0 +1,475 @@
+  rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base,
+  rclcpp::node_interfaces::NodeGraphInterface::SharedPtr node_graph,
+  rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr node_logging,
+  const std::string & action_name,
```
include `<string>`

ahcorde commented on rclcpp_action/src/generic_client.cpp:
```diff
@@ -0,0 +1,475 @@
+  const Goal goal, size_t goal_size, const SendGoalOptions & options)
+{
+  // uuid + goal
+  size_t goal_request_size = sizeof(rclcpp_action::GoalUUID) + goal_size;
```
include `<cstddef>`

ahcorde commented on rclcpp_action/src/generic_client.cpp:
```diff
@@ -0,0 +1,475 @@
+  // uuid + goal
+  size_t goal_request_size = sizeof(rclcpp_action::GoalUUID) + goal_size;
+  std::shared_ptr<uint8_t> goal_request_msg(
+    new uint8_t[goal_request_size], std::default_delete<uint8_t[]>());
```
include `<cstdint>`

ahcorde commented on rclcpp_action/src/generic_client.cpp:
```diff
@@ -0,0 +1,475 @@
+    reinterpret_cast<rclcpp_action::GoalUUID *>(goal_request_msg.get());
+  *uuid = this->generate_goal_id();
+
+  std::memcpy(
```
include `<cstring>`

ahcorde commented on rclcpp_action/src/generic_client.cpp:
```diff
@@ -0,0 +1,475 @@
+  return async_send_goal(static_cast<void *>(goal_request_msg.get()), options);
+}
+
+std::shared_future<typename GenericClientGoalHandle::SharedPtr>
```
include `<future>`

ahcorde commented on rclcpp_action/src/generic_client.cpp:
```diff
@@ -0,0 +1,475 @@
+  // To prevent the list from growing out of control, forget about any goals
+  // with no more user references
+  {
+    std::lock_guard<std::recursive_mutex> guard(goal_handles_mutex_);
```
include `<mutex>`

ahcorde commented on rclcpp_action/src/generic_client.cpp:
```diff
@@ -0,0 +1,475 @@
+GenericClient::async_cancel_all_goals(CancelCallback cancel_callback)
+{
+  auto cancel_request = std::make_shared<CancelRequest>();
+  std::fill(
```
include `<algorithm>`

ahcorde commented on rclcpp_action/src/generic_client_goal_handle.cpp:
```diff
@@ -0,0 +1,155 @@
+  return info_.stamp;
+}
+
+std::shared_future<typename GenericClientGoalHandle::WrappedResult>
```
include `<future>`

ahcorde commented on rclcpp_action/src/generic_client_goal_handle.cpp:
```diff
@@ -0,0 +1,155 @@
+std::shared_future<typename GenericClientGoalHandle::WrappedResult>
+GenericClientGoalHandle::async_get_result()
+{
+  std::lock_guard<std::recursive_mutex> guard(handle_mutex_);
```
include `<mutex>`

ahcorde commented on rclcpp_action/src/generic_client_goal_handle.cpp:
```diff
@@ -0,0 +1,155 @@
+GenericClientGoalHandle::set_result(const WrappedResult & wrapped_result)
+{
+  std::lock_guard<std::recursive_mutex> guard(handle_mutex_);
+  status_ = static_cast<int8_t>(wrapped_result.code);
```
include `<cstdint>`

ahcorde commented on rclcpp_action/test/test_generic_client.cpp:
```diff
@@ -0,0 +1,1164 @@
+            feedback_message.feedback.sequence.push_back(1);
+            feedback_publisher->publish(feedback_message);
+            client_executor.spin_once();
+            for (size_t i = 1; i < static_cast<size_t>(goal_request->goal.order); ++i) {
```
include `<cstddef>`

ahcorde commented on rclcpp_action/test/test_generic_client.cpp:
```diff
@@ -0,0 +1,1164 @@
+        ActionCancelGoalResponse::SharedPtr response)
+      {
+        rclcpp_action::GoalUUID zero_uuid;
+        std::fill(zero_uuid.begin(), zero_uuid.end(), 0u);
```
include `<algorithm>`

ahcorde commented on rclcpp_action/test/test_generic_client.cpp:
```diff
@@ -0,0 +1,1164 @@
+  template<typename FutureT>
+  void dual_spin_until_future_complete(std::shared_future<FutureT> & future)
+  {
+    std::future_status status;
```
include `<future>`

ahcorde commented on rclcpp_action/test/test_generic_client.cpp:
```diff
@@ -0,0 +1,1164 @@
+  const char * const namespace_name{"/rclcpp_action/test/client"};
+  const char * const action_name{"fibonacci_test"};
+
+  std::map<
```
include `<map>`

ahcorde commented on rclcpp_action/test/test_generic_client.cpp:
```diff
@@ -0,0 +1,1164 @@
+
+  std::map<
+    rclcpp_action::GoalUUID,
+    std::pair<
```
include `<utility>`

Barry-Xu-2018 committed:
Add the missing header files and address other review comments

Signed-off-by: Barry Xu <barry.xu@sony.com>

Barry-Xu-2018 commented:
@ahcorde   

Thank you for your review.  
I have addressed all your review comments at https://github.com/ros2/rclcpp/pull/2759/commits/9f366f0131c155db9beb1eee6af0e9a6ef134066.

ahcorde commented:
Pulls: ros2/rclcpp#2759
Gist: https://gist.githubusercontent.com/ahcorde/222e1cc850a365d120f7b68ee1ba7ed4/raw/6509370844cf61fc69fe2ce57683a00cbedb620e/ros2.repos
BUILD args:  --packages-above-and-dependencies rclcpp_action
TEST args:  --packages-above rclcpp_action
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/15295
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22647)](http://ci.ros2.org/job/ci_linux/22647/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16889)](http://ci.ros2.org/job/ci_linux-aarch64/16889/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=2163)](http://ci.ros2.org/job/ci_linux-rhel/2163/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23457)](http://ci.ros2.org/job/ci_windows/23457/)

fujitatomoya commented:
@ahcorde thanks for reviewing and running CI.

@Barry-Xu-2018 windows test build failed, can you take a look?

fujitatomoya commented on rclcpp_action/include/rclcpp_action/generic_client.hpp:
```diff
@@ -0,0 +1,323 @@
+  std::shared_future<typename GenericClientGoalHandle::SharedPtr>
+  async_send_goal(
+    const Goal goal,
+    size_t goal_size,
```
checking my understanding here.

this is required because it needs to allocate goal uuid following this goal message, right? message constructor does not generate the uuid but action client does internally. so that it needs to know the goal content size from the application, and copy it the actually message field to `rcl`?

generic service client does not need to have this `size` argument, since it just can be passed to `rcl` with `void *` pointer?

Barry-Xu-2018 replied:
@fujitatomoya 

> this is required because it needs to allocate goal uuid following this goal message, right? message constructor does not generate the uuid but action client does internally. so that it needs to know the goal content size from the application, and copy it the actually message field to rcl?

Yes, your understanding is correct. 
It might be possible to infer the size of an instantiated Goal type through introspection information, but that process would likely be very complex. On the other hand, the memory block pointed to by the pointer comes from the user, who can more conveniently provide the size of this memory block.  

> generic service client does not need to have this size argument, since it just can be passed to rcl with void * pointer?

Yes. It does not need to combine the data being sent.

fujitatomoya reviewed (approved):
lgtm, just one thing i would like to confirm.

Barry-Xu-2018 committed:
Add the missing visibility control macros

Signed-off-by: Barry Xu <barry.xu@sony.com>

Barry-Xu-2018 commented:
>  windows test build failed, can you take a look?

I fixed it by https://github.com/ros2/rclcpp/pull/2759/commits/3f155e8597c7760fb44033224dc07992a7f7ad18.   
However, there seems to be an issue with the CI system, and currently, I am unable to log in to rerun Window tests

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22655)](http://ci.ros2.org/job/ci_linux/22655/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=16897)](http://ci.ros2.org/job/ci_linux-aarch64/16897/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=2171)](http://ci.ros2.org/job/ci_linux-rhel/2171/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23480)](http://ci.ros2.org/job/ci_windows/23480/)

fujitatomoya commented:
@ahcorde thanks for running CI. all your comments are addressed, could you approve?
